src/validator.rs

```diff
@@ -959,17 +917,49 @@ impl<'a> Validator<'a> {
     }
 }
 
+pub fn types_match(node: &Node, other: &Node) -> HowserResult<bool> {
+    trace!("types_match");
+    let node_getter = node.capabilities
```

Comment:
Many, this capabilities business really needs to go. With this incidental complexity gone, I bet the problem will start to look a lot simpler.